sdk/textanalytics/ai-text-analytics/src/models.ts

```diff
@@ -0,0 +1,443 @@
+  "EntityRecognition",
+  "KeyPhraseExtraction",
+  "PiiEntityRecognition",
+  "LanguageDetection",
+  "SentimentAnalysis",
+] as const;
+
+/**
+ * Type of actions supported by the {@link TextAnalysisClient.analyze} method.
+ */
+// eslint-disable-next-line @typescript-eslint/no-redeclare
+export type AnalyzeActionName = typeof AnalyzeActionName[number];
+
+/**
+ * The type of parameters for every action in ${@link AnalyzeActionName}.
+ */
+export type AnalyzeActionParameters<ActionName extends AnalyzeActionName> =
+  ActionName extends "EntityLinking"
+    ? EntityLinkingAction
+    : ActionName extends "EntityRecognition"
+    ? EntityRecognitionAction
+    : ActionName extends "KeyPhraseExtraction"
+    ? KeyPhraseExtractionAction
+    : ActionName extends "PiiEntityRecognition"
+    ? PiiEntityRecognitionAction
+    : ActionName extends "SentimentAnalysis"
+    ? SentimentAnalysisAction
+    : ActionName extends "LanguageDetection"
+    ? LanguageDetectionAction
+    : never;
+
+/**
+ * The type of results of every action in ${@link AnalyzeActionName}.
+ */
+export type AnalyzeResult<ActionName extends AnalyzeActionName> = ActionName extends "EntityLinking"
+  ? EntityLinkingResult[]
+  : ActionName extends "EntityRecognition"
+  ? EntityRecognitionResult[]
+  : ActionName extends "KeyPhraseExtraction"
+  ? KeyPhraseExtractionResult[]
+  : ActionName extends "PiiEntityRecognition"
+  ? PiiEntityRecognitionResult[]
+  : ActionName extends "SentimentAnalysis"
+  ? SentimentAnalysisResult[]
+  : ActionName extends "LanguageDetection"
+  ? LanguageDetectionResult[]
+  : never;
+
+/**
+ * Enum of possible error codes of a {@link TextAnalysisError}.
+ */
+export const KnownTextAnalysisErrorCode = { ...KnownErrorCode, ...KnownInnerErrorCode };
+
+/**
+ * Type describing an API error.
+ */
+export interface TextAnalysisError {
+  /**
+   * A code describing the kind of error produced. See {@link KnownTextAnalysisErrorCode}.
+   */
+  readonly code: string;
+  /**
+   * A message from the service explaining the error
+   */
+  readonly message: string;
+  /**
+   * The target of the particular error (for example, the name of an invalid parameter)
+   */
+  readonly target?: string;
+}
+
+/**
+ * Base type for results of an action corresponding to a single input document.
+ */
+export interface TextAnalysisSuccessResult {
+  /**
+   * Unique, non-empty document identifier.
+   */
+  readonly id: string;
+
+  /**
+   * Statistics about the input document and how it was processed by the service.
+   * This property will have a value when includeStatistics is set to true in
+   * the client call.
+   */
+  readonly statistics?: TextDocumentStatistics;
+
+  /**
+   * An array of warning data corresponding to this document.
+   *
+   * If no warnings were returned, this array will be empty.
+   */
+  readonly warnings: DocumentWarning[];
+
+  /**
+   * Discriminant to determine if this is an error result.
+   */
+  readonly error?: undefined;
```

Comment:
curious: double optional?

Reply:
@witemple-msft do you know why this can not be `readonly error: undefined;`?

Reply:
The difference between `error?: undefined` and `error: undefined` is that the latter one will enforce that the property  _is present_ on object literals of that type, so `error?: undefined` is what we want in this case because we want to say that it's either _not there_, or if it is, it's `undefined`.

Kind of a weird case, but this ends up being the discriminant tag for these result types. By checking `result.error` the compiler will know whether it was a success or error result.